algorithmic-strategies/dynamic-programming.md

```diff
@@ -0,0 +1,302 @@
+_Fig. Fibonacci of 5. Notice how the same subproblems are repeatedly called!_
+
+Instead of solving the same problems over and over again we can solve these problems by storing them in a _memo_ and using the stored subproblems to make calculating the larger problem more efficient.
+
+Let's take a look at a few variations of `fibonacci` which make use of memoization to improve their performance.
+
+<br />
+
+<details>
+
+<summary>Forward iterative solution. Starts from the base cases of [0, 1] building up to the desired number.</summary>
+
+```python
+def fibonacci(n):
+    if n == 0 or n == 1:
+        return n
+
+    solutions = [0, 1]
+    current = 2
+
+    while current <= n:
+        solutions.append(solutions[current - 1] + solutions[current - 2])
+        current += 1
+
+    return solutions[n]
+```
+
+</details>
+
+<details>
+
+<summary>Forward recursive solution. A recursive adaptation of the forward iterative solution.</summary>
+
+```python
+def fibonacci_recursive(n, solutions=None, current=None):
+    if solutions is None:
+        solutions = [0, 1]
+        current = 2
+
+    solutions.append(solutions[current - 1] + solutions[current - 2])
+
+    if n <= current:
+        return solutions[n]
+
+    return fibonacci_recursive(n, solutions, current + 1)
+```
+
+</details>
+
+<details>
+
+<summary>Traditional recursive solution. Based on the typical branching recursive solution, but memoizes calculations as they are performed.</summary>
+
+```python
+def fibonacci_recursive(n, solutions=None):
+    if solutions is None:
+        solutions = {}
+
+    if n in solutions:
+        return solutions[n]
+
+    if n == 0 or n == 1:
+        solutions[n] = n
+    else:
+        solutions[n] = (fibonacci_recursive(n - 1, solutions) +
+            fibonacci_recursive(n - 2, solutions))
+
+    return solutions[n]
+```
+
+</details>
+
+To really appreciate the improvements that dynamic programming can bring, we can try running `fibonacci(40)` with the original, recursive implementation and note how long this takes to run. If we then run _any_ of the versions that use dynamic-programming techniques, we should see a significant improvement!
+
+### !callout-info
+
+## Fibonacci Challenges
+
+Notice how when solving `fibonacci(n)` we only need `fibonacci(n-1)` and `fibonacci(n-2)`. Using this observation, can we reduce our space complexity from \(O(n)\) to \(O(1)\) for the forward iterative solution? Would a similar approach improve the forward recursive solution?
```

Comment:
took out the forward references and rephrased the suggestion a bit to still try to be specific about to which recursive solution we are referring.

Reply:
merging these two suggestions, since GH won't let me batch apply them at the same time
```suggestion
Notice how when solving `fibonacci(n)`, the only solutions we need are `fibonacci(n-1)` and `fibonacci(n-2)`. Using this observation, can we reduce our space complexity from \(O(n)\) to \(O(1)\) for the iterative solution? Are we able to apply this strategy to the recursive version of that solution?
```